File: NeetCode/Arrays & Hashing/Top K Frequent Elements.js
class Solution {
    /**
     * Time Complexity:O(N + M log M), where N is the length of nums and M is the number of distinct elements.
     * Space Complexity: O(M) for sorting entries 
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */
    topKFrequent(nums, k) {
        const HashMap = new Map();

        for (let i = 0; i< nums.length; i++){
            if(HashMap.has(nums[i])){
                HashMap.set(nums[i], HashMap.get(nums[i])+1);
            } else{
                HashMap.set(nums[i], 1);
            }
        }
        let sorted = [...HashMap].sort((a, b) => b[1] - a[1]);

        const topK = [];

        for (let i = 0; i < k; i++) {
            topK.push(sorted[i][0]);
        return topK;
        }
    }
}

File: Leetcode/Top-K-Frequent-Elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counter = Counter(nums) # create hashmap 
        buckets = [[] for _ in range(len(nums) + 1)] # create empty list for each index
       
        # # insert hashmap into buckets (bucket sort) freqency : [values]
        for value, freq in counter.items():
            buckets[freq].append(value)
        
        # return to k
        result = []
        for n in range(len(buckets) - 1, 0 , -1): # move backwards
            for i in buckets[n]:
                result.append(i)
                if len(result) == k:
                    return result
    

        # Time Complexity: O(n)
        # Space Complexity: O(n)